### PR TITLE
Enhance null check for post parameters to cater for cache expiry #2972

### LINKED ISSUE #2972: Enhance null check for post parameters to cater for cache expiry
In action classes,
`Assumption.assertNotNull(....`
to be replaced with
`Assumption.assertPostParamNotNull(...`

It has been done in some actions, but need to be done in the remaining actions.

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #2972

### CODE REVIEW

**Comment by sujeet14108** [comment]
Ready for review :)

File: src/main/java/teammates/ui/automated/FeedbackResponseAdjustmentWorkerAction.java [hunk 1]
```diff
@@ -36,13 +36,13 @@ protected String getActionMessage() {
     @Override
     public void execute() {
         String courseId = getRequestParamValue(ParamsNames.COURSE_ID);
-        Assumption.assertNotNull(courseId);
+        Assumption.assertPostParamNotNull(ParamsNames.COURSE_ID, courseId);
 
         String sessionName = getRequestParamValue(ParamsNames.FEEDBACK_SESSION_NAME);
-        Assumption.assertNotNull(sessionName);
+        Assumption.assertPostParamNotNull(ParamsNames.COURSE_ID, sessionName);
```
Comment (leeyimin) [inline src/main/java/teammates/ui/automated/FeedbackResponseAdjustmentWorkerAction.java]: Wrong param name here!

File: src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java [hunk 2]
```diff
@@ -34,9 +34,9 @@ protected ActionResult execute() throws EntityDoesNotExistException {
 
         String moderatedQuestionId = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_QUESTION_ID);
 
-        Assumption.assertNotNull(Const.StatusMessages.NULL_POST_PARAMETER_MESSAGE, courseId);
-        Assumption.assertNotNull(Const.StatusMessages.NULL_POST_PARAMETER_MESSAGE, feedbackSessionName);
-        Assumption.assertNotNull(Const.StatusMessages.NULL_POST_PARAMETER_MESSAGE, instructorUnderModerationEmail);
+        Assumption.assertPostParamNotNull(Const.StatusMessages.NULL_POST_PARAMETER_MESSAGE, courseId);
+        Assumption.assertPostParamNotNull(Const.StatusMessages.NULL_POST_PARAMETER_MESSAGE, feedbackSessionName);
+        Assumption.assertPostParamNotNull(Const.StatusMessages.NULL_POST_PARAMETER_MESSAGE, instructorUnderModerationEmail);
```
Comment (leeyimin) [inline src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java]: It may be better to standardise using parameter name as the first argument to this method, instead of using the status message.

File: src/main/java/teammates/ui/controller/InstructorFeedbackCopyAction.java [hunk 3]
```diff
@@ -21,10 +21,10 @@ protected ActionResult execute() throws EntityDoesNotExistException {
...
         String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
 
-        Assumption.assertNotNull(copiedFeedbackSessionName);
-        Assumption.assertNotNull(copiedCourseId);
-        Assumption.assertNotNull(courseId);
-        Assumption.assertNotNull(feedbackSessionName);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.COPIED_FEEDBACK_SESSION_NAME, copiedFeedbackSessionName);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.COPIED_COURSE_ID, copiedCourseId);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_SESSION_NAME, courseId);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.COURSE_ID, feedbackSessionName);
```
Comment (leeyimin) [inline src/main/java/teammates/ui/controller/InstructorFeedbackCopyAction.java]: Lines 26, 27 seems to have their parameter names swapped.

File: src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveAction.java [hunk 4]
```diff
@@ -305,24 +305,24 @@ private FeedbackResponseAttributes extractFeedbackResponseData(
...
+        Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_SESSION_NAME, response.feedbackSessionName);
 
         response.courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
-        Assumption.assertNotNull("Null feedback courseId", response.courseId);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.COURSE_ID, response.courseId);
 
         response.feedbackQuestionId = getRequestParamValue(
                 Const.ParamsNames.FEEDBACK_QUESTION_ID + "-" + questionIndx);
-        Assumption.assertNotNull("Null feedbackQuestionId", response.feedbackQuestionId);
+        Assumption.assertPostParamNotNull("Null feedbackQuestionId", response.feedbackQuestionId);
```
Comment (leeyimin) [inline src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveAction.java]: It may be better to use the full param name `Const.ParamsNames.FEEDBACK_QUESTION_ID + "-" + questionIndx` to standardise with how the method is used throughout *Actions.

File: src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveAction.java [hunk 5]
```diff
@@ -305,24 +305,24 @@ private FeedbackResponseAttributes extractFeedbackResponseData(
...
         Assumption.assertEquals("feedbackQuestionId Mismatch", feedbackQuestionAttributes.getId(),
                                 response.feedbackQuestionId);
 
         response.recipient = getRequestParamValue(
                 Const.ParamsNames.FEEDBACK_RESPONSE_RECIPIENT + "-" + questionIndx + "-" + responseIndx);
-        Assumption.assertNotNull("Null feedback recipientEmail", response.recipient);
+        Assumption.assertPostParamNotNull("Null feedback recipientEmail", response.recipient);
```
Comment (leeyimin) [inline src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveAction.java]: Can consider using the full param name `Const.ParamsNames.FEEDBACK_RESPONSE_RECIPIENT + "-" + questionIndx + "-" + responseIndx`.

File: src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveAction.java [hunk 6]
```diff
@@ -305,24 +305,24 @@ private FeedbackResponseAttributes extractFeedbackResponseData(
...
 
         String feedbackQuestionType = getRequestParamValue(
                 Const.ParamsNames.FEEDBACK_QUESTION_TYPE + "-" + questionIndx);
-        Assumption.assertNotNull("Null feedbackQuestionType", feedbackQuestionType);
+        Assumption.assertPostParamNotNull("Null feedbackQuestionType", feedbackQuestionType);
```
Comment (leeyimin) [inline src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveAction.java]: Can consider using the full param name `Const.ParamsNames.FEEDBACK_QUESTION_TYPE + "-" + questionIndx`.

**Comment by sujeet14108** [comment]
@leeyimin 
In Action.java
` regkey = getRegkeyFromRequest();
Assumption.assertNotNull(regkey);`
Do we need to change` assertNotNull `to `assertPostParamNotNull`  in this also as it is not post parameter ?

**Comment by sujeet14108** [comment]
Also in `InstructorFeedbackPublishAction` test is failing if  null check for post parameters

`Assumption.assertPostParamNotNull(Const.ParamsNames.COURSE_ID, courseId);`

Same problem in `InstructorFeedbackUnpublishAction`

**Comment by leeyimin** [comment]
For InstructorFeedbackPublishAction, you need to update the test for empty post parameters from line 74 to 81, possibly like this.
```
try {
    publishAction.executeAndPostProcess();
    signalFailureToDetectException("NullPostParameterException expected");
} catch (NullPostParameterException e) {
    errorMessage = e.getMessage();
}

assertEquals(String.format(Const.StatusCodes.NULL_POST_PARAMETER, ParamsNames.COURSE_ID), errorMessage);
```
You need to do the same for the test case for empty feedback session name and the InstructorFeedbackUnpublishActionTest.

**Comment by leeyimin** [comment]
For `regkey` in Action, my thought was that the name of the method implies that it was getting the value from the post parameters and it does indeed uses `getRequestParamValue`. So it seems to make sense to also change `assertNotNull` to `assertPostParamNotNull` for it. If you think this is not so clear from the method name, may be it is okay to leave this one out.

**Comment by sujeet14108** [comment]
@leeyimin You are right about the `regkey`

**Comment by sujeet14108** [comment]
@leeyimin Done :+1:
